api/envoy/extensions/transport_sockets/tls/v3/common.proto

```diff
@@ -484,10 +488,13 @@ message CertificateValidationContext {
   // certificate chain will be subject to validation by :ref:`CRL <envoy_v3_api_field_extensions.transport_sockets.tls.v3.CertificateValidationContext.crl>`.
   bool only_verify_leaf_cert_crl = 14;
 
-  // Config for the max number of intermediate certificates in chain that are parsed during verification.
-  // This does not include the leaf certificate. If configured, and the certificate chain is longer than allowed, the certificates
-  // above the limit are ignored, and certificate validation will fail. The default limit is 100,
-  // though this can be system-dependent.
-  // https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_verify_depth.html
+
+  // Defines maximum depth of a certificate chain accepted in verification. This number does not include the leaf, so a depth of 1 allows the leaf and one CA certificate.
+  // If a trusted issuer appears in the chain, but in a depth larger than configured, the certificate validation will fail.
+  // See `BoringSSL SSL_CTX_set_verify_depth <https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#SSL_CTX_set_verify_depth>`
+  // If you use OpenSSL, its behavior is different from BoringSSL, this will define a limit on the number of certificates between the end-entity and trust-anchor certificates.
```

Comment:
emm...I wasn't sure about this since it passes a long time, I'm a little forget about the background. need to double check

Reply:
I just found this difference today, the comment mixed boringssl behavior and openssl behavior before. They are not the same
BoringSSL:
This number does not include the leaf

OpenSSL:
Neither the end-entity nor the trust-anchor certificates count against depth. If the certificate chain needed to reach a trusted issuer is longer than depth+2

you can check the behavior of testcase https://github.com/envoyproxy/envoy/pull/22350/files#diff-f86b77f38f9ce6f2295761c59e534cb90cf8c095454db7106e6d66cd87e209d3R133, it should fail with openssl but succeed with boringssl.

Reply:
have you actually test that on the openssl?

If yes, there are probably two things need to think about. One is the test you added is going to break https://github.com/envoyproxy/envoy-openssl. The second is whether we want to expose different behavior in the API, maybe we can let the OpenSSL match the same behavior like it will process the depth as `max_depth+1`.

But I don't know whether those things should be processed in the Envoy main repo, or it should be done at Envoy OpenSSL repo. Those probably need maintainer help here.

Reply:
No, I didn't test it with openssl. I just copy the official documentation here. @ggreenway I need your help, how should we do for openssl?

Reply:
I'm not sure if this is the same issue or not, but is this due to https://bugs.chromium.org/p/boringssl/issues/detail?id=533 (which refers to https://github.com/openssl/openssl/commit/0daccd4dc1f1ac62181738a91714f35472e50f3c)?

I wouldn't block this change on the potential to break envoy-openssl.

Reply:
It seems not due the issue https://bugs.chromium.org/p/boringssl/issues/detail?id=533.

The issue here is the 'depth' has different means in boringssl and openssl,

in boringssl, only leaf cert not count against depth. There are test cases verifying its behavior.
in openssl, neither leaf cert nor trust-anchor count against depth. I did not test it, just copy the openssl documentation to let users aware of the difference. Is that enough?

Reply:
@ggreenway ping.

Reply:
Yes, I think documenting the difference is good enough.